### ISSUE TITLE
Research URI / Deep linking approach

### ISSUE THREAD

**Issue by AmmarHumackicSQA** [near-projects/wallet-selector#50]
Research about deep linking for wallet integration.

https://github.com/near/NEPs/pull/10#issuecomment-1000442297

Example: 

If an end-user wants to transfer NEAR, the URI path would be something like:

near-wallet://action/Transfer/10?to=mike.near"

**Comment by lewis-sqa** [comment]
Some background investigation into the MetaMask mobile app:

- Uses a package called `react-native-batch` to handle deep links. This is actually something React Navigation (the dominant navigation framework in the eco-system) does for you: https://reactnavigation.org/docs/deep-linking/#set-up-with-bare-react-native-projects. Behind the scenes, they both use the Linking module: https://reactnative.dev/docs/linking.
- It works by subscribing to incoming urls and funnels it through a parser (which is also applicable to universal links). It’s then up to the app to determine how it should respond to the url which usually results in navigating to a particular route and forwarding the parameters.
- It should be noted that deep linking does pose a security risk as it opens up an attack vector for sending code directly into an app. As always, external input should be considered dangerous and steps to sanitise the data is advisable.

**Comment by lewis-sqa** [comment]
Some background investigation into the [near-wallet](https://github.com/near/near-wallet) project:

- The URL constants mostly limit aspects of the UI such as hiding nav links and/or footer. 
- These limitations are mostly driven by a specific reducer called [`flowLimitation`](https://github.com/near/near-wallet/blob/13a7dc5b5d5c2206e102643a9d1b9129fdcfd095/packages/frontend/src/redux/slices/flowLimitation/index.js)

**Comment by lewis-sqa** [comment]
My limited knowledge in the domain would suggest capitalising on the already established structure of a [Transaction](https://nomicon.io/RuntimeSpec/Transactions.html). As we know, Transactions support [action batching](https://nomicon.io/RuntimeSpec/Transactions.html#batched-transaction) in their nature which would address Mike’s 2nd additional use case.

Here is my initial proposal of a URL:

```
{scheme}://transaction?value={base64-encoded-transaction-contents}
```

I’m not completely sold on the parameter `value` so it might be that we move a level down:

```
{scheme}://transaction?signer_id={account-id}&public_key={public-key}&nonce={nonce}&receiver_id={account-id}&block_hash={hash}&actions={base64-encoded-actions}
```

It’s also noted that historic paths might be something we want to support during migration such as `login`. New specialised use cases such as offline signing will simply add another path in the mix such as:

```
{scheme}://sign?key1={value1}&key2={value2}
```

Ultimately, the focus from my understanding will be the successful approval of a new NEP that will outline the different URL paths (to address the various use cases) and the shape of their respective query parameters.

**Comment by lewis-sqa** [comment]
We can use the [`popstate`](https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event) event to track when URL changes occur in the browser.

**Comment by mikedotexe** [comment]
> Ultimately, the focus from my understanding will be the successful approval of a new NEP that will outline the different URL paths (to address the various use cases) and the shape of their respective query parameters.

Thanks, @lewis-sqa for the responses and it seems you understand this clearly. I think the task here is to make some sane judgements on how we route, and not necessarily be tied to historic routes from the NEAR Wallet.

We want to have an NEP at some point for sure. I think my main intent on creating this ticket is so we have a consistent pathing schema on Wallet Selector side. We can have "shims" for wallets that don't abide by that.

In case it's helpful, I'll explain more:
As we've discussed with Daryl, there needs to be a tight integration with `near-api-js` ([repo](https://github.com/near/near-api-js)) and the Wallet Selector.

Wallet Selector will (likely) instantiate a connection, providing which wallet the has selected, for instance `SenderWallet` of class `Injected`. Then the user pokes around the dApp and clicks a button to sign and send a transaction. The Wallet Selector + `near-api-js` says, "hey, use the route `/transfer` for user `mike.near` for whichever wallet the user selected"
Let's now say that SenderWallet didn't implement a `/transfer` route and instead implemented `/send`. At that point, I believe we'd inform SenderWallet that we'd all like to use `/transfer` but in the meantime we'd have a shim redirecting to the route that it understands.

Basically, we need to attempt to use canonical paths for transactions, and be able to reroute where needed.

**Comment by lewis-sqa** [comment]
It looks like there are two problem sets in this spike:

1. Implement a consistent framework to define Deep Links for consumers such websites and QR codes. This will make it easy for apps to ingest these URLs and navigate the user to the correct flows.
2. Enable a tighter integration with `near-api-js` to align RPC endpoints?

I believe we can easily solve problem 1 with a similar approach to Ethereum by providing an NPM package that enforces an NEP spec (TBD) for both parsing and building compliant URLs.

My understanding of `near-api-js` is rather limited so I'm currently unable to provide a solution here. I think at this point we need to move ahead to user stories with clear acceptance criteria to fulfil the use cases here.

Without a clearly defined goal in this spike, I'm going to go ahead and close this in an effort to avoid it staying open indefinitely. Once an NEP spec has been raised and approved, we can raise a story to create a separate package that addresses problem 1 similar to [`eth-url-parser`](https://www.npmjs.com/package/eth-url-parser).
